# MÓDULO 25 — Operações Básicas SQL (Respostas com Código)



Cada questão traz **o código usado** e o **resultado** logo abaixo.

In [10]:
import pandas as pd

# Carregar base (separador ';')
df = pd.read_csv("G:\Meu Drive\Programação\Ciencia de dados\Aprofundamento de Analytics\Operacoes basicas SQL\TB_VENDAS_TAREFA.csv", sep=";", encoding="utf-8")
df.columns = [c.strip().upper() for c in df.columns]

# Tipos e coluna auxiliar
df["VALOR_UNID"] = pd.to_numeric(df["VALOR_UNID"], errors="coerce")
df["UNIDADES"] = pd.to_numeric(df["UNIDADES"], errors="coerce")
df["_VALOR_TOTAL"] = df["VALOR_UNID"] * df["UNIDADES"]

df.head(3)

ID_COMPRA  ID_CLIENTE   PRODUTO  VALOR_UNID  UNIDADES  _VALOR_TOTAL
0       1247        9081  CAMISETA         NaN         2           NaN
1       1248        9081     CALCA         NaN         1           NaN
2       1250        9560  CAMISETA         NaN         1           NaN

## 1) Faça uma consulta que retorne os nomes dos produtos distintos que temos na base de venda.

In [11]:
# Produtos distintos (ordenados alfabeticamente)
produtos_distintos = sorted(df["PRODUTO"].dropna().unique().tolist())
produtos_distintos

['BERMUDA', 'CALCA', 'CAMISETA', 'MEIA', 'VESTIDO']

**Resultado:**

- BERMUDA
- CALCA
- CAMISETA
- MEIA
- VESTIDO

## 2) Faça uma consulta que retorne a contagem dos clientes distintos que temos na nossa base, não esqueça de renomear sua coluna.

In [12]:
# Contagem de clientes distintos (coluna renomeada)
clientes_distintos = df["ID_CLIENTE"].nunique()
import pandas as pd
pd.DataFrame({"CLIENTES_DISTINTOS": [clientes_distintos]})

CLIENTES_DISTINTOS
0                  22

**Resultado:**

| CLIENTES_DISTINTOS |
| --- |
| 22 |

## 3) Faça uma consulta que retorne uma coluna com os produtos distintos e o valor_unid de cada produto. Porém apenas para produtos onde o valor_unid é maior ou igual a 50 reais.

In [13]:
# Produtos com VALOR_UNID >= 50 (distintos)
res = (df.loc[df["VALOR_UNID"] >= 50, ["PRODUTO", "VALOR_UNID"]]
         .dropna()
         .drop_duplicates()
         .sort_values(["VALOR_UNID", "PRODUTO"], ascending=[False, True]))
res

PRODUTO  VALOR_UNID
5  VESTIDO       102.0

**Resultado:**

| PRODUTO | VALOR_UNID |
| --- | --- |
| VESTIDO | 102.00 |

## [DESAFIO] 4) Faça uma consulta que retorne o ID das compras e o valor total gasto nessa compra (Valor_unidade X Unidades) das 5 compras com mais valor total gasto.

In [14]:
# Top 5 compras por valor total
res = (df[["ID_COMPRA","VALOR_UNID","UNIDADES","_VALOR_TOTAL"]]
         .sort_values("_VALOR_TOTAL", ascending=False)
         .head(5)
         .reset_index(drop=True)
         .rename(columns={"_VALOR_TOTAL":"VALOR_TOTAL"}))
res

ID_COMPRA  VALOR_UNID  UNIDADES  VALOR_TOTAL
0       1267       102.0         2        204.0
1       1252       102.0         1        102.0
2       1256       102.0         1        102.0
3       1247         NaN         2          NaN
4       1248         NaN         1          NaN

**Resultado:**

| ID_COMPRA | VALOR_UNID | UNIDADES | VALOR_TOTAL |
| --- | --- | --- | --- |
| 1267 | 102.00 | 2 | 204.00 |
| 1252 | 102.00 | 1 | 102.00 |
| 1256 | 102.00 | 1 | 102.00 |
| 1247 | nan | 2 | nan |
| 1248 | nan | 1 | nan |

## 5) Faça uma consulta que retorne os produtos e a média do preço da unidade dos produtos, ordenando do maior para o menor.

In [15]:
# Média do preço da unidade por produto (ordem desc)
res = (df.groupby("PRODUTO", as_index=False)["VALOR_UNID"]
         .mean()
         .rename(columns={"VALOR_UNID":"MEDIA_VALOR_UNID"})
         .sort_values("MEDIA_VALOR_UNID", ascending=False))
res

PRODUTO  MEDIA_VALOR_UNID
4   VESTIDO             102.0
0   BERMUDA               NaN
1     CALCA               NaN
2  CAMISETA               NaN
3      MEIA               NaN

**Resultado:**

| PRODUTO | MEDIA_VALOR_UNID |
| --- | --- |
| VESTIDO | 102.00 |
| BERMUDA | nan |
| CALCA | nan |
| CAMISETA | nan |
| MEIA | nan |